File: gnoboard/ios/gnoboard/Sources/Bridge/GoBridge.m
#import <React/RCTBridgeModule.h>


@interface RCT_EXTERN_MODULE(GoBridge, NSObject)

RCT_EXTERN_METHOD(initBridge:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject);

RCT_EXTERN_METHOD(closeBridge:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject);

RCT_EXTERN_METHOD(listKeyInfo:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject);

RCT_EXTERN_METHOD(createAccount:(NSString *)nameOrBech32
                  mnemonic:(NSString *)mnemonic
                  bip39Passwd:(NSString *)bip39Passwd
                  password:(NSString *)password
                  account:(nonnull NSNumber *)account
                  index:(nonnull NSNumber *)index
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject);

RCT_EXTERN_METHOD(selectAccount:(NSString *)nameOrBech32
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject);

RCT_EXTERN_METHOD(call:(NSString *)packagePath
                  fnc:(NSString *)fnc
                  args:(NSArray *)args
                  gasFee:(NSString *)gasFee
                  gasWanted:(nonnull NSNumber *)gasWanted
                  password:(NSString *)password
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject);

RCT_EXTERN_METHOD(exportJsonConfig:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject);

+ (BOOL)requiresMainQueueSetup
{
  return NO;  // only do this if your module initialization relies on calling UIKit!
}

- (dispatch_queue_t)methodQueue
{
  return dispatch_queue_create("com.facebook.React.AsyncLocalStorageQueue", DISPATCH_QUEUE_SERIAL);
}

@end

File: gnoboard/src/native_modules/GoBridge.ts
import { GoBridgeInterface } from "./GoBridgeInterface";

class NoopGoBridge implements GoBridgeInterface {
  initBridge() {
    return Promise.reject();
  }

  closeBridge() {
    return Promise.reject();
  }

  setPassword(_password: string): Promise<void> {
    return Promise.reject();
  }

  generateRecoveryPhrase(): Promise<string> {
    return Promise.reject();
  }

  listKeyInfo(): Promise<Array<Object>> {
    return Promise.reject();
  }

  createAccount(
    _nameOrBech32: string,
    _mnemonic: string,
    _bip39Passw: string,
    _password: string,
    _account: Number,
    _index: Number,
  ): Promise<Object> {
    return Promise.reject();
  }

  selectAccount(_nameOrBech32: string): Promise<Object> {
    return Promise.reject();
  }

  getActiveAccount(): Promise<Object> {
    return Promise.reject();
  }

  query(
    _path: string,
    _data_b64: string,
  ): Promise<string> {
    return Promise.reject();
  }

  call(
    _packagePath: string,
    _fnc: string,
    _args: Array<string>,
    _gasFee: string,
    _gasWanted: Number,
  ): Promise<string> {
    return Promise.reject();
  }

  exportJsonConfig() {
    return Promise.reject();
  }
}

export const GoBridge: GoBridgeInterface = new NoopGoBridge();
